src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                // argument expression.
+                OptimizeTemporaries(actualArguments, storesToTemps, temporariesBuilder);
+            }
 
             // Step two: If we have a params array, build the array and fill in the argument.
             if (expanded)
             {
-                actualArguments[actualArguments.Length - 1] = BuildParamsArray(syntax, methodOrIndexer, argsToParamsOpt, rewrittenArguments, parameters, actualArguments[actualArguments.Length - 1]);
+                var array = VisitAndBuildParamsArray(syntax, methodOrIndexer, argsToParamsOpt, arguments, parameters, actualArguments[actualArguments.Length - 1], argumentsAreVisited);
+
+                if (argumentsWillBeReused)
+                {
+                    BoundLocal boundTemp = _factory.StoreToTemp(array, out BoundAssignmentOperator store);
+                    actualArguments[^1] = boundTemp;
+                    storesToTemps.Add(store);
+                }
+                else
+                {
+                    actualArguments[^1] = array;
+                }
+            }
+
+
+            if (argumentsWillBeReused)
+            {
+                // We can't optimize away the temporaries if arguments are going to be reused,
+                // because we don't know whether a specific index is going to be reused or not
+                // (and therefore re-evaluated).
+                foreach (BoundExpression expression in storesToTemps)
+                {
+                    Debug.Assert(expression is BoundSequence or BoundAssignmentOperator);
+                    if (expression is BoundAssignmentOperator assignment)
+                    {
+                        temporariesBuilder.Add(((BoundLocal)assignment.Left).LocalSymbol);
+                    }
+                }
+
+                if (incomingStores != null)
+                {
+                    // We're in the middle of a compound assignment, it will handle the sequence
+                    // of stores for us. Just add the stores to the builder we were given.
+                    incomingStores.AddRange(storesToTemps);
+                    storesToTemps.Free();
+                }
+                else
+                {
+                    // If this was an interpolated string conversion that needed a receiver, the
+                    // first store is actually the receiver, not the argument, so turn the receiver
+                    // into a sequence. Otherwise, turn the first actual argument into that sequence.
+
+#pragma warning disable CS8619 // https://github.com/dotnet/roslyn/issues/54153
+                    ref BoundExpression resultLocation = ref interpolatedStringHandlerArgumentIndex > -1 && requiresInstanceReceiver
+                        ? ref visitedReceiver!
+                        : ref actualArguments[0];
+#pragma warning restore CS8619
+
+                    resultLocation = _factory.Sequence(
```

Comment:
>resultLocation = _factory.Sequence( [](http://example.com/codeflow?start=20&length=35)

Are we testing scenarios when the receiver is a structure:
- an LValue
- an RValue

Wineed to make sure the right value is used for the receiver and across all iterpolations in the arguments

Reply:
I believe I've added tests covering these scenarios.

Reply:
> I believe I've added tests covering these scenarios.

I'd like us to follow up offline to make sure we are on the same page about what I think we should test.